lib/BaseSlackHandler.js

```diff
@@ -0,0 +1,172 @@
+"use strict";
+
+const rp = require('request-promise');
+const Promise = require('bluebird');
+const promiseWhile = require("./promiseWhile");
+const getSlackFileUrl = require("./substitutions").getSlackFileUrl;
+
+/**
+ * @constructor
+ * @param {Main} main the toplevel bridge instance through which to
+ * communicate with matrix.
+ */
+function BaseSlackHandler(main) {
+    this._main = main;
+}
+
+BaseSlackHandler.prototype.replaceChannelIdsWithNames = function (message, token) {
+    var main = this._main;
+
+    // match all channelIds
+    var testForName = message.text.match(/<#(\w+)\|?\w*?>/g);
+    var iteration = 0;
+    var matches = 0;
+    if (testForName && testForName.length) {
+        matches = testForName.length;
+    }
+    return promiseWhile(function () {
+        // Do this until there are no more channel ID matches
+        return iteration < matches;
+    }, function () {
+        // foreach channelId, pull out the ID
+        // (if this is an emote msg, the format is <#ID|name>, but in normal msgs it's just <#ID>
+        var id = testForName[iteration].match(/<#(\w+)\|?\w*?>/)[1];
+        var channelsInfoApiParams = {
+            uri: 'https://slack.com/api/channels.info',
+            qs: {
+                token: token,
+                channel: id
+            },
+            json: true
+        };
+        main.incRemoteCallCounter("channels.info");
+        return rp(channelsInfoApiParams).then((response) => {
+            if (response && response.channel && response.channel.name) {
+                console.log("channels.info: " + id + " mapped to " + response.channel.name);
+                message.text = message.text.replace(/<#(\w+)\|?\w*?>/, "#" + response.channel.name);
```

Comment:
I'm afraid you're going to have to elaborate here, We are checking in the if statement it is successful? Otherwise we log and move on?

Reply:
Given a string with A, B, and C to be replaced in it the block here will iterate over the information that was previously grepped from the string. This results in 3 requests for information, of which any could fail. If B's request were to fail, but C was successful then the replacement string becomes `A, C, <not changed>` rather than `A, <not changed>, C` (assuming the string was `A, B, C`).